### DISCUSSION TITLE
Weather the line number(relative number) is supported?

### DISCUSSION THREAD

**Discussion by charlielin** [sxyazi/yazi#704] [1. Q&A]
I didn't find any configs about line number

**Comment by rrveex** [comment]
I put this in ~/.config/yazi/init.lua

```
function Current:render(area)
	self.area = area

	local markers = {}
	local items = {}

	local cursor = cx.active.current.cursor
	local offset = cx.active.current.offset

	for i, f in ipairs(Folder:by_kind(Folder.CURRENT).window) do
		local name = Folder:highlighted_name(f)

		local idx = i - 1 - cursor + offset
		if idx == 0 then idx = cursor end

		local item = ui.ListItem(ui.Line { ui.Span(string.format("%3d", idx)), Folder:icon(f), table.unpack(name) })
		-- Highlight hovered file
		if f:is_hovered() then
			item = item:style(THEME.manager.hovered)
		else
			item = item:style(f:style())
		end
		items[#items + 1] = item

		-- Mark yanked/selected files
		local yanked = f:is_yanked()
		if yanked ~= 0 then
			markers[#markers + 1] = { i, yanked }
		elseif f:is_selected() then
			markers[#markers + 1] = { i, 3 }
		end
	end
	return ya.flat { ui.List(area, items), Folder:linemode(area), Folder:markers(area, markers) }
end
```
Looks like this:
![image](https://github.com/sxyazi/yazi/assets/91850165/71a1a4ee-b229-437a-a241-8550a49c6e6e)
![image](https://github.com/sxyazi/yazi/assets/91850165/1fa11a26-b42e-44f1-9601-c92de8c4ff41)

The "selected" / "yanked" markers overlap the numbers, that could be fixed with something greater than `%3d` in `format`. Whatever. I just discovered yazi today :)

Next I'll try "relative up/down"... Should work witn `0..1` keys mapped to a plugin which keeps state (and thus remembers the "count" of lines to jump to); then send a message with `move` and `count` on up/down back. Have to play a bit more to figure stuff out.

**Comment by sxyazi** [comment]
Here is a plugin for relative line number navigation, https://github.com/dedukun/relative-motions.yazi.

It is still a work in progress, but is basically functional, just lacks some documentation and line number display. I'll be working on adding these to the plugin in the next day or two.

**Comment by rrveex** [comment]
@sxyazi I don't understand how the plugin you mention keeps the count... or is it supposed to be called something like `plugin motions --args='13k'`? (this wouldn't make sense of course, since you could just bind `arrow 13` to some key and wouldn't need a plugin)

I ended up with this:
```
local nums = {
	['1'] = 1, ['2'] = 2, ['3'] = 3, ['4'] = 4, ['5'] = 5, ['6'] = 6, ['7'] = 7, ['8'] = 8, ['9'] = 9, ['0'] = 0
}

return {
	entry = function(_, args)
		local num = nums[args[1]]
		if num == nil then
			if MOVE_LINES == 0 then MOVE_LINES = 1 end
			if args[1] == 'up' then
				ya.manager_emit("arrow", { -MOVE_LINES })
			elseif args[1] == 'dn' then
				ya.manager_emit("arrow", { MOVE_LINES })
			end
			MOVE_LINES = 0
		else
			MOVE_LINES = MOVE_LINES * 10 + num
		end
	end,
}
```
With `MOVE_LINES` global in ~/.config/init.lua because on `v0.2.3` my yazi doesn't keep the state (tried example from "sync context" / test - `i` stays `0`). And I can't really use `main` because it never finishes "preload mime" or something (the state example works, though).

**Reply by sxyazi** [reply]
Only sync plugins have `state` - you can use it in the `entry` function, like `entry = function(state)`. For async plugins, please use the `state` from `ya.sync(function(state) end)`, see https://yazi-rs.github.io/docs/plugins/overview#async-context

**Comment by sxyazi** [comment]
Hi guys, https://github.com/dedukun/relative-motions.yazi, this plugin is now available 🎉

See https://github.com/sxyazi/yazi/issues/642#issuecomment-1962214771

**Comment by rrveex** [comment]
Hi @sxyazi , thanks!

Unfortunately, I still can't run `main`. Updated cargo, pulled last commits (`8ea7556`), built release/debug, still the same problem:

![image](https://github.com/sxyazi/yazi/assets/91850165/690fcbf2-66ad-4dba-b506-ce98ab1d4f5d)

The upper instance is started from the yazi git directory; the lower one from home. The lower one stays with "preloading mime" forever.

When I move one file down in the upper one (from dir `assets` to file `Cargo.lock`), it looks like this:

![image](https://github.com/sxyazi/yazi/assets/91850165/b1f63a03-0a4b-487a-9f61-29356c52a474)

I have run this with no config folder at all, deleted `~/.local/state/yazi` before running.

Log (same for both paths I start from, it keeps saying this):

```
ERROR yazi::lives::lives: runtime error: [string "yazi-plugin/src/loader.rs:59:28"]:18: attempt to index a nil value (local 'style')
stack traceback:
	[C]: in metamethod 'index'
	[string "yazi-plugin/src/loader.rs:59:28"]:18: in function 'folder.peek'
    at yazi-fm/src/lives/lives.rs:66
```

**Reply by lpnh** [reply]
Hey, hi! It seems that the last commit has introduced some unexpected behavior. I've experienced something similar after updating the repo today. What I did was build it using the previous commit, and everything is working again.

Here's what I did after cloning the current repo:

```
git checkout e51e8ad

cargo clean

cargo build --release
```

After that, I simply removed the previous binary and copied the new one to my path.

**Reply by sxyazi** [reply]
Hi, could you zip your `~/.config/yazi` and upload it to let me see what happened?

**Reply by rrveex** [reply]
As I wrote before, I deleted (moved away) the ~/.config/yazi before trying and reporting the problem.

**Reply by rrveex** [reply]
https://github.com/sxyazi/yazi/assets/91850165/9652cee1-28dd-493c-8611-d9ac328c0167

**Reply by sxyazi** [reply]
Ah I see. Please open a bug report, I will respond to you there efficiently.

**Comment by rrveex** [comment]
Thank you Filipe! `e51e8ad` works better, looks OK when starting in the yazi repo and navigating inside there. 

Navigating to other dirs though (home, /etc, whatever) still triggers the "preloader mime" problem.

**Comment by sxyazi** [comment]
> Navigating to other dirs though (home, /etc, whatever) still triggers the "preloader mime" problem

This has been fixed in https://github.com/sxyazi/yazi/issues/722, thanks for reporting @rrveex !
